enzyme/Enzyme/MLIR/Analysis/ActivityAnnotations.cpp

```diff
@@ -0,0 +1,1319 @@
+  // StringRef calleeName = cast<SymbolRefAttr>(call.getCallableForCallee())
+  //                            .getLeafReference()
+  //                            .getValue();
+
+  ChangeResult changed = ChangeResult::NoChange;
+  // Unify the value origin summary with the actual lattices of function
+  // arguments
+  // Collect the origins of the function arguments, then collect the alias
+  // classes of the destinations
+  auto *p2sets = getOrCreateFor<PointsToSets>(call, call);
+  SmallVector<ValueOriginSet> argumentOrigins;
+  SmallVector<AliasClassSet> argumentClasses;
+  for (auto &&[i, argOperand] : llvm::enumerate(call.getArgOperands())) {
+    // Value origin might be sparse, might be dense
+    ValueOriginSet argOrigins;
+    auto *argClasses = getOrCreateFor<AliasClassLattice>(call, argOperand);
+    if (argClasses->isUndefined()) {
+      // Not a pointer, use the sparse lattice state
+      auto *sparseOrigins =
+          getOrCreateFor<ForwardOriginsLattice>(call, argOperand);
+      (void)argOrigins.join(sparseOrigins->getOriginsObject());
+    } else {
+      // Unify all the origins
+      // Since we're not keeping track of argument depth, we need to union the
+      // arg origins with everything it points to.
+      traversePointsToSets(argClasses->getAliasClassesObject(), *p2sets,
+                           [&](DistinctAttr aliasClass) {
+                             (void)argOrigins.join(
+                                 before.getOrigins(aliasClass));
+                           });
+    }
+    argumentClasses.push_back(argClasses->getAliasClassesObject());
+    argumentOrigins.push_back(argOrigins);
+  }
+
+  // TODO: Does the traversal order matter here?
+  for (const auto &[destClass, sourceOrigins] : summary) {
+    ValueOriginSet callerOrigins;
+    for (Attribute sourceOrigin : sourceOrigins.getElements()) {
+      unsigned argNumber =
+          cast<ArgumentOriginAttr>(sourceOrigin).getArgNumber();
+      (void)callerOrigins.join(argumentOrigins[argNumber]);
+    }
+
+    AliasClassSet callerDestClasses;
+    if (auto pseudoClass = dyn_cast_if_present<PseudoAliasClassAttr>(
+            destClass.getReferencedAttr())) {
+      // Traverse the points-to sets.
+      AliasClassSet current = argumentClasses[pseudoClass.getArgNumber()];
+      unsigned depth = pseudoClass.getDepth();
+      while (depth > 0) {
+        AliasClassSet next;
+        if (current.isUndefined()) {
+          // TODO: what should be done here?
+          // Activity annotations requires converged pointer info. If we have
+          // incomplete points-to info, we can't currently tell if it's because
+          // the points-to info hasn't _yet_ been computed (in which case we
+          // bail out here expecting to be called again with more complete
+          // points-to info), or if the points-to info has converged, this
+          // signifies reading from uninitialized memory.
```

Comment:
Leverage undefined behavior. The compiler assumes undefined behavior, such as reading from uninitialized memory, never happens.